primitives/trie/src/cache/mod.rs

```diff
@@ -36,43 +36,285 @@
 
 use crate::{Error, NodeCodec};
 use hash_db::Hasher;
-use hashbrown::HashSet;
 use nohash_hasher::BuildNoHashHasher;
-use parking_lot::{Mutex, MutexGuard, RwLockReadGuard};
-use shared_cache::{SharedValueCache, ValueCacheKey};
+use parking_lot::{Mutex, MutexGuard};
+use schnellru::LruMap;
+use shared_cache::{ValueCacheKey, ValueCacheRef};
 use std::{
-	collections::{hash_map::Entry as MapEntry, HashMap},
-	sync::Arc,
+	collections::HashMap,
+	sync::{
+		atomic::{AtomicU64, Ordering},
+		Arc,
+	},
+	time::Duration,
 };
 use trie_db::{node::NodeOwned, CachedValue};
 
 mod shared_cache;
 
 pub use shared_cache::SharedTrieCache;
 
-use self::shared_cache::{SharedTrieCacheInner, ValueCacheKeyHash};
+use self::shared_cache::ValueCacheKeyHash;
 
 const LOG_TARGET: &str = "trie-cache";
 
-/// The size of the cache.
+/// The maximum amount of time we'll wait trying to acquire the shared cache lock
+/// when the local cache is dropped and synchronized with the share cache.
+///
+/// This is just a failsafe; normally this should never trigger.
+const SHARED_CACHE_WRITE_LOCK_TIMEOUT: Duration = Duration::from_millis(100);
+
+/// The maximum number of existing keys in the shared cache that a single local cache
+/// can promote to the front of the LRU cache in one go.
+///
+/// If we have a big shared cache and the local cache hits all of those keys we don't
+/// want to spend forever bumping all of them.
+const SHARED_NODE_CACHE_MAX_PROMOTED_KEYS: u32 = 1792;
```

Comment:
I picked it empirically by basically just running a node with the stats logs enabled, looking at the numbers as new blocks are coming in, and I just picked a number that's not too big while still being larger enough to not hit it.

We probably could further fine-tune it and pick a better value (as I didn't spend a ton of time on this); ideally we'd probably want to make it configurable? That could be useful to make it possible to experiment with different values and see how it can affect performance, but that can be done in a future PR.